Rewrite of ProtocolConformance.md (#2)
dabrahams/SwiftRef: ProtocolConformance.md

> # 1 Protocol Conformance

Reviewer: We need to choose a column limit.  I had wrapped at 80, I'd be happy with 100 if that works better for you, but we should agree on a number.

Reviewer: You should integrate this material into the chapter list as part of the PR.

Author: Thanks, Dave.  

I'm not used to writing prose with hard returns at the end of each line, so I've been a bit sloppy and undisciplined.  As I've been working on the Statements chapter, I've been getting much better at that.  Let's stick with 80.

Question about wrapping:  When the content ultimately is rendered for final use in a wider or narrower format than the selected , how does one deal with the out-of-place hard returns? 

Thank you for the substantive comments.  I will review them in detail, and revise accordingly.

Reviewer: What editor are you using?  You might be able to set something up to wrap paragraphs when you save.

Reviewer: The hard returns don't create a paragraph break unless you turn on [GFM](https://github.github.com/gfm/) in your Jekyll config

Author: Thanks.  I'm using Sublime Text.  I've turned on the ruler, and set it at 80.  I think that will be enough to keep me on track.

When a named concrete type[^1] `T` is conformed to a protocol `P`, a set of implementations--*witnesses*--is 
determined, one for each requirement of `P`.   That set of witnesses is referred to as the protocol conformance
of `T: P`.  This document specifies how Swift determines a conformance.

> [^1]: What about tuples and [SE-0283](https://github.com/apple/swift-evolution/blob/master/proposals/0283-tuples-are-equatable-comparable-hashable.md)?

Reviewer: There's a problem with the phrasing; there is no defined “when” for this.  We have three compilation phases to work with for talking about “when”: tokenization, parsing, and translation.

```suggestion
In a given scope, a concrete type `T` **conforms** to a protocol `P` if and only if:
- `T` is a nominal type and the conformance is implied [<-- link to section on how this implication is
  determined] by a
  *[type-inheritance-clause](https://docs.swift.org/swift-book/ReferenceManual/Types.html#grammar_type-inheritance-clause)* 
  visible in that scope.
- `T` is is a non-nominal type that implicitly conforms to `P` as described in [section dealing
  with tuples].

A protocol **conformance** of `T` to `P` is a triple (`T`, `P`, *M*), where *M* is a mapping from
each requirement *r* of `P` onto a corresponding **witness** that satisfies [<--link to requirement satisfaction] *r* for `T`.   

| Requirement | Witness |
|-|-|
| associated type | concrete type nested in `T` |
| non-`static` method | non-`static`, non-`class` method of `T` |
| non-operator `static` method | `static` or `class` method of `T`, or an `enum` case as described in [section describing that horrible new feature] |
| static operator | `static` or `class` operator of `T`, or a concrete module-scope operator |

This chapter describes how a conformance's mapping from requirements to witnesses is determined.
```

Author: I like that approach.

For your consideration, I'd like to suggest:
- a `typealias` (declared or inferred) is the witness for an `associatedtype`
- for requirements, we should distinguish between `func` and `var` requirements
- we could use the term instance rather than non-`static`

Reviewer: * typealiases really *are* aliases for types, rather than types themselves.  I think the witness is the actual type being aliased.  The typealias is just the thing that connects the witness to the conforming type.
* Ah yes, I left out properties.  And subscripts.  The witnesses are are actually getter functions or getter/setter pairs. I'lll write some text for that once we get this a little more settled.  We should drive this toward a state where it can be landed in master even if incomplete.
* “instance” works for me.
* I notice I also left out generic method /subscript requirements, whose instances are actually generics.

## 1.1 Creation of a Conformance
A **protocol conformance** is the set of witnesses used by a concrete type to satisfy the requirements of a 
protocol.  The syntactic declaration, `T: P`, that type `T` conforms to protocol `P` is separate and distinct 
from the determination of the set of witnesses that is a conformance.  The substance of a conformance is not 
declared.  It is created in response to a type being declared to conform to a protocol.  The set of witnesses 

> that is a conformance is inferred from the context.

Reviewer: Suggest dropping this paragraph.  The term was defined in my suggestion above.  There may not be any syntactic declaration:

```swift
struct X : Hashable { } // Implies Equatable
```

> For a type `T` to conform to a protocol `P`, `T` must be declared to conform to `P`, and `T` must have at 

Reviewer: Not true for tuples

> least one implementation for each protocol requirement of `P`.  A named concrete type is declared to conform to 

Reviewer: Well, "have" is poorly defined here.  The implementations may come from protocol extensions or be derived via associated type inference, etc.

Can we use the term “nominal type" please?  After all, this is named but not nominal:

```swift
typealias Named = (Int, Int)
```

a protocol in one of two ways.  A non-generic concrete type is expressly declared to conform to a protocol.  A 
concretization of a generic type is implicitly declared to conform to a protocol based on the pattern established 

> by the generic type's declaration and extensions.

Reviewer: I think everything of value in this paragraph may be implied by my suggestion above?

> ><sub>***Conretization of a Generic Type**

Reviewer: Per the description at https://www.w3schools.com/tags/tag_sub.asp I can't understand why you'd want the subscript tag on this whole paragraph.

>A concretization of a generic type is a specialized version of the generic type formed by replacing the generic 

>  type's type parameters with concrete type arguments.  A generic type may specify that its concretizations will 

Reviewer: It's great that you wrote it down, but I think the definition of concretization probably belongs in https://dabrahams.github.io/SwiftRef/chapters/1600%20Generics.html

Reviewer: We can't really talk about the generic type “specifying” anything.  The program may contain a  conformance declaration for a generic type.

I'm a little uncomfortable with the term “pattern” here, but I'm not entirely sure why so let it ride for now.

 conform, or conditionally may conform, to one or more protocols, which establishes a pattern for protocol 
 conformance.  When a concretization is formed, its protocol conformances are determined by that pattern.*</sub>

> A declaration that `T` conforms to `P` further constitutes, with respect to each protocol `O`*<sub>i</sub>* which 

Reviewer: ```suggestion
A declaration that `T` conforms to `P` further constitutes, with respect to each protocol `O`*<sub>i</sub>* that 
```

`P` refines, a declaration that `T` conforms to `O`*<sub>i</sub>*.  Thus, for the declaration of `T: P` to be 
valid, the implicit declarations of each `T: O`*<sub>i</sub>* must be valid.  For each conformance `T: O`*<sub>i</sub>*, 
a distinct set of protocol witnesses is established. 

> A type `T` cannot be declared to conform to a protocol `P` if, within the visible scope, another declaration 
> exists of `T: P`.  This rule holds true even where competing declarations are conditional with disjoint conditions.  

Reviewer: Well, that's only true if it's another *explicit* declaration of conformance.

The way to say this is:

```suggestion
If, in a given scope, more than one conformance declaration of a given type to a
given protocol is visible, the program is ill-formed.  This rule holds true even where
 competing declarations are conditional with disjoint conditions.  
```

Perhaps an example here.

> In a given scope, a type can conform to a protocol in only one way.

Reviewer: I don't think that sentence adds anything.

## 1.2 Requirements

> A **protocol requirement** *m* is a statement in the declaration of a protocol that a type declared to conform 

Reviewer: I think we need a top-level chapter called “Protocols” for this material.

to the protocol must have a member satisfying *m*.

If one protocol refines another protocol, the requirements of the latter are not part of the requirements of 
the former.  The requirements of a protocol are not overridden by a protocol that refines the protocol.  Thus, 

> the conformances formed by a refining protocol do not include witnesses for requirements of the refined protocol.

Reviewer: You said this 3 ways.  I'm not sure it *needs* to be said at all because it will be implied by a strict reading of the rest of the document, though I understand why you've included it: it was a point of confusion during discussions.  The right thing to do with stuff like that is to put it in a *note:*, which makes it “non-normative.”  I think it's important to separate the language *required* for understanding from the annotations and other supporting material.

## 1.3 Implementations

Given `T: P` and a protocol requirement *m* of `P`, an **implementation** of *m* is any member of `T` that satisfies 

> *m*.  A type may have more than one implementation of a requirement.

Reviewer: I don't think we want to define “implementation,” I think we should be talking about “witness candidates” for a given requirement.  When you call something an implementation of a requirement it implies some ongoing connection between the candidate and the requirement, but no such relationship exists.

It doesn't need to be a member, in the case of operators.
